feat(hex): use registry v2 for packages rather than hex.pm http API

bryannaegele opened the pull request:
## Changes

- use hex registry v2 for package information
- use hex.pm package API for supplemental package metadata when package is hosted there.
- support `registryAliases` for private repo/registry mapping
- run `mix.repo add` in mix manager for private repos

## Context

https://github.com/renovatebot/renovate/issues/29622

This changes the `hex` datasource to use hex registry v2 for package management. Hex registries are not required to implement an http API like `hex.pm` has, so the prior implementation prevented updating of packages in private repos such as oban pro. Additionally, due to the way mix works when updating dependencies, `mix.lock` files could not be updated when such a dependency existed and the mix manager ran `mix deps.update <<updated packages>>`. 

In order to keep as much parity with the prior implementation, hex.pm is used for supplemental package metadata where possible.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

bryannaegele commented:
Leaving this as a draft until there's an answer for this https://github.com/renovatebot/renovate/issues/29622#issuecomment-2177278783

I am no TS expert, so please feel free to jump in and fix up whatever can be done better.

viceice commented on lib/modules/datasource/hex/package.spec.ts:
```diff
@@ -0,0 +1,17 @@
+// import { mockDeep } from 'jest-mock-extended';
```
remove

viceice commented on lib/modules/datasource/hex/package.spec.ts:
```diff
@@ -0,0 +1,17 @@
+// import { mockDeep } from 'jest-mock-extended';
+import { gunzipSync } from 'zlib';
```
use async functions

viceice commented on lib/modules/manager/mix/artifacts.ts:
```diff
@@ -11,8 +11,8 @@ import {
 import { regEx } from '../../../util/regex';
-
 import type { UpdateArtifact, UpdateArtifactsResult } from '../types';
+import { aliasRecordToRepositories } from './utils';
```
split datasource and manager changes to separate PR

bryannaegele replied:
#29775

viceice commented on lib/modules/datasource/hex/signed.ts:
```diff
@@ -0,0 +1,154 @@
+// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
```
this should be generated on build and not checked into git. or fix it to a readable format for proper linting.

bryannaegele replied:
Can you help with how you would go about that? I am not a TS developer and this is what was generated by https://www.npmjs.com/package/ts-proto#usage, unedited in any way. There are a lot of options that can be passed to it but it's all foreign to me. The comment and eslint ignore are output by ts-proto.

I'm not sure how generating the files on build ends up with a different result. It would still be the same code, just lint ignored? And the code has to be there for unit tests, so I'm not clear on how a build step improves things.

viceice replied:
you need to add that package to dev dependencies, then add a generate step to package.json scripts. there are already some generation steps as samples.

viceice reviewed (commented):
reduce fixtures, we don't need all versions for testing, ~3 are enough. ensure the tgz file is at minimum too

bryannaegele commented:
> reduce fixtures, we don't need all versions for testing, ~3 are enough. ensure the tgz file is at minimum too

Is this strictly necessary? The tgz has the registry version of the package wrapped in a signed wrapper. I don't have any way to create a version with less releases in it that would have a valid signature.

I can try to find a different package as an example but this is one of the rare ones with retirements.

bryannaegele commented:
Mix changes split to #29775

bryannaegele commented:
Ok. I think everything has been addressed. This and #29775 are ready to go @viceice and @rarkins

viceice commented on .github/workflows/build.yml:
```diff
@@ -207,6 +207,9 @@ jobs:
           node-version: ${{ needs.setup-build.outputs.node-version }}
           os: ${{ runner.os }}
 
+      - name: Generate protos
+        run: pnpm create-protos
+
```
revert those changes and do it in generate script, see package.json for other generate scripts

bryannaegele replied:
That was my first approach. I tried doing it in generate but that creates a dependency issue and requires modifying a number of test commands and likely the github workflows to run generate where they currently do not. One example is linting which does not have a dependency on `setup` currently. This causes the files to not be present when there are references to them.

<img width="547" alt="image" src="https://github.com/user-attachments/assets/baf36835-009d-4f76-8289-de3ee5aa96e6">

Introducing a generic protos creation step was the least intrusive solution I could get to.

If you can assist in getting to a working solution with generate, it would be much appreciated.

viceice commented on package.json:
```diff
@@ -8,14 +8,15 @@
     "compile:ts": "tsc -p tsconfig.app.json",
     "config-validator": "ts-node lib/config-validator.ts",
     "create-json-schema": "ts-node tools/generate-schema.ts && prettier --write --cache 'renovate-schema.json'",
+    "create-protos": "ts-node tools/generate-protos.ts",
```
```suggestion
    "generate:create-protos": "ts-node tools/generate-protos.ts",
```
then all other changes can be reverted

bryannaegele replied:
See note in the other comment about generate not running in the lint pipeline.

viceice commented on tools/generate-protos.ts:
```diff
@@ -0,0 +1,98 @@
+  process.exit(-1);
+});
+
+const moveFiles = async (sourceDir: string, destDir: string): Promise<void> => {
```
use normal function declaration

viceice commented on tools/generate-protos.ts:
```diff
@@ -0,0 +1,98 @@
+  }
+})();
+
+function istanbulIgnoreFile(file_path: string): string {
+  fs.appendFile(file_path, '\n// istanbul ignore file', function (err) {
+    if (err) {
+      throw err;
+    }
+  });
+
+  return '';
+}
```
exclude in jest.config.ts instead

viceice commented on tools/generate-protos.ts:
```diff
@@ -0,0 +1,98 @@
+  await generateProto(protos_path, 'signed.proto');
+  istanbulIgnoreFile(`${protos_path}/package.ts`);
+  istanbulIgnoreFile(`${protos_path}/signed.ts`);
+  await moveFiles(
+    `${process.cwd()}/lib/modules/datasource/hex/protos`,
+    `${process.cwd()}/lib/modules/datasource/hex`,
+  );
```
why moving? why not directly generate on right path?

bryannaegele replied:
It is only able to generate the file to the current directory where the proto is located. Currently, the proto files are separated for organizational purposes so it would just mean having the proto files interspersed with the ts files, e.g. `signed.proto` and `signed.ts` in the same directory. Either approach works so happy to go whichever way you'd like.

bryannaegele commented:
Ok. I am at a point now where trying to get the protos to generate on build in `generate:` and putting them to the correct place for the sake of not committing the generated files is at a breaking point for me. There is a dependency chain issue here that I cannot find a way resolve without extensive amount of changes to the workflows and/or package scripts.

* the generated files must live in the datasource, so `lib/modules/datasource/hex` to compile
* the `generate:` command runs on a sparse checkout that does not include `lib`
* adding that path to the sparse checkout results in error:
```. prepare: generating hashes
. prepare: ERROR: ENOENT: no such file or directory, scandir 'lib/modules/manager'```
* subsequent steps such as lint do a full checkout
* any generated file would be overwritten by the full checkout, so generation would have to happen again after every step that does that

I understand the desire to not add the `create-protos` step I added which was working fine [here](https://github.com/renovatebot/renovate/pull/29756/commits/a433ab7e0604ddc7c004f2b7e969d9c42272e7c4), but trying to shoehorn it into `generate:` is infeasible imo. I've attempted it two times now and it just doesn't work. It also can't go into `build:` like the docs because `compile` runs before that and the generated files must be there.

If the files must be generated and not committed, can we please use the new workflow I had introduced?

I feel the simplest solution is still to commit the generated files which we're ignoring for coverage and linting in jest config now anyways and documenting the instructions for generating them should they ever need to be updated. It would be great to reconsider that approach again.

Thanks

viceice commented:
> Ok. I am at a point now where trying to get the protos to generate on build in `generate:` and putting them to the correct place for the sake of not committing the generated files is at a breaking point for me. There is a dependency chain issue here that I cannot find a way resolve without extensive amount of changes to the workflows and/or package scripts.
> 
> * the generated files must live in the datasource, so `lib/modules/datasource/hex` to compile
> * the `generate:` command runs on a sparse checkout that does not include `lib`
> * adding that path to the sparse checkout results in error:
> ```. prepare: generating hashes
> . prepare: ERROR: ENOENT: no such file or directory, scandir 'lib/modules/manager'```
> * subsequent steps such as lint do a full checkout
> * any generated file would be overwritten by the full checkout, so generation would have to happen again after every step that does that
> 
> I understand the desire to not add the `create-protos` step I added which was working fine [here](https://github.com/renovatebot/renovate/pull/29756/commits/a433ab7e0604ddc7c004f2b7e969d9c42272e7c4), but trying to shoehorn it into `generate:` is infeasible imo. I've attempted it two times now and it just doesn't work. It also can't go into `build:` like the docs because `compile` runs before that and the generated files must be there.
> 
> If the files must be generated and not committed, can we please use the new workflow I had introduced?
> 
> I feel the simplest solution is still to commit the generated files which we're ignoring for coverage and linting in jest config now anyways and documenting the instructions for generating them should they ever need to be updated. It would be great to reconsider that approach again.
> 
> Thanks

@zharinov can you help with fixing the CI in an elegant way?

zharinov commented:
Since we don't expect generated files to change frequently, why not commit them? At the CI level we could check the generated code is identical to the committed one, and well... sometimes we have to replace it manually.

I don't know how how bad consequences are, but if we ignore `signed.ts` and `package.ts`, our typescript code is broken unless the `pnpm install` command is executed.

viceice commented:
> Since we don't expect generated files to change frequently, why not commit them? At the CI level we could check the generated code is identical to the committed one, and well... sometimes we have to replace it manually.
> 
> I don't know how bad consequences are (probably, the language server will suffer), but if we ignore `signed.ts` and `package.ts`, our typescript project is broken unless the `pnpm install` command is executed.

it's already broken without install

zharinov commented:
> it's already broken without install

Well, fair enough. I think generate phase may just quit early if there is no `hex` subdirectory, and it should avoid usage of `utils/exec` in favor of the Node's basic one.

viceice commented:
> > it's already broken without install
> 
> Well, fair enough. I think generate phase may just quit early if there is no `hex` subdirectory, and it should avoid usage of `utils/exec` in favor of the Node's basic one.

sounds good. only use node build-in or deps, but no code from src inside prepare steps?

zharinov commented:
Yes, because if the `lib/` directory is absent for this particular CI task, the module import attempt will fail.

bryannaegele commented:
If it helps on frequency of update with regard to committing the generated code, these protos are [5 and 6 years old](https://github.com/hexpm/specifications/tree/main/registry) at this point. They may never be updated again. I suppose the generated code may need an update at some point due to changes in TS or a bug but the code seems pretty simple, so low risk there.

viceice commented:
ok, how big are the generated files? maybe we can check them in and ignore in linting

viceice commented:
or move the proto files to a different root folder which is checked out

bryannaegele commented:
> ok, how big are the generated files? maybe we can check them in and ignore in linting

| file | generated | formatted | formatted and encoding functions manually deleted |
| --- | --- | --- | --- |
| signed.ts | 132 LOC | 155 LOC | 119 LOC |
| package.ts | 570 LOC | 638 LOC | 479 LOC |

zharinov commented:
> ok, how big are the generated files? maybe we can check them in and ignore in linting

Yeah, I think having the task like `codegen:hex-datasource-protobuf` is enough

bryannaegele commented:
> or move the proto files to a different root folder which is checked out
Missed this comment. The issue isn't really where the proto files live, it's that the generated files have to exist in the hex datasource folder and can't be overwritten by a subsequent checkout that a lot of the steps do. So that would introduce having to generate and copy the files from a different location anytime a checkout happens in CI.

bryannaegele commented:
@zharinov and @viceice - have we landed on just committing the generated files or is someone taking over to get codegen how y'all want it?

bryannaegele commented:
Thanks. I'll get that done.

One thing to note is the base functionality has not changed, it's all enhancements, so it shouldn't be breaking. And we probably want to get #29775 reviewed to be merged in conjunction with this one. They are dependent on each other.

bryannaegele commented:
Alright. Made the following requested changes:
- remove codegen from CI and setup
- commit generated files
- add readme with info on how to generate the proto files
- renamed script name in package per @zharinov's suggestion

I have kept the generation script in tools so that if anyone should need to regenerate the files, it's as simple as possible.

esambo commented:
@bryannaegele FYI, you have merge conflicts

bryannaegele commented:
> @bryannaegele FYI, you have merge conflicts

Are y'all ready to merge this and the other PR? I'd like to not keep resolving the conflicts unless it's going to get merged.

viceice commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+          const { payload: payload } = Signed.decode(signedPackage);
+          const registryPackage = Package.decode(payload);
+
+          const releases: Release[] = registryPackage.releases.map(
```
use for-of loop

viceice commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+    } else if (packageName.startsWith('repo:')) {
+      const [, repoName, hexPackageName] = packageName.split(':');
+      urlPath = `/packages/${hexPackageName}`;
+      this.hexRepoName = repoName;
```
no, we've concurrent calls, don't store state in datasource . use local function variable

viceice commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+      if (this.hexRepoName === 'hexpm' && releaseResult.releases.length > 0) {
+        const metadataUrl = joinUrlParts(this.hexAPIBaseUrl, urlPath);
+
+        logger.debug(`Package metadata url: ${metadataUrl}`);
```
```suggestion
        logger.trace(`Package metadata url: ${metadataUrl}`);
```

viceice commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+          .getJsonSafe(metadataUrl, HexAPIPackageMetadata)
+          .unwrap();
+
+        // istanbul ignore if
```
no, add test

viceice commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+          releaseResult.sourceUrl = packageMetadata.meta?.links.Github;
+          releaseResult.homepage = packageMetadata.html_url;
+
+          const releasesWithMeta = releaseResult.releases.map((rel) => {
```
use for-of loop

viceice commented on tools/generate-hex-protos.ts:
```diff
@@ -0,0 +1,86 @@
+function generateProto(protos_path: string, file: string): Promise<string> {
+  return new Promise((resolve, reject) => {
+    console.log(`Current directory: ${process.cwd()}`);
+    fs.readdirSync(process.cwd()).forEach((file) => {
```
use `fs/promises` api

bryannaegele commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -1,28 +1,33 @@
-  override readonly releaseTimestampSupport = true;
-  override readonly releaseTimestampNote =
-    'The release timestamp is determined the `inserted_at` field in the results.';
+  override readonly releaseTimestampSupport = false;
```
Timestamps are not a piece of metadata in the registry metadata. They are only a piece of metadata that hex.pm supports via its HTTP API which is not a requirement to run a registry.

[Available fields](https://github.com/hexpm/specifications/blob/main/registry/package.proto).

`inserted_at` field [is preserved](https://github.com/renovatebot/renovate/pull/29756/files#diff-99307b00eaeb9c01bb1a10333bde71e11f13a5c3eab4c19a1c1101726d4f1d37R18) but that field is only available for a package served from hexpm.

Can you explain why it's a showstopper? I understand it's a potentially breaking change but the original implementation was technically incorrect. I'm not sure how the `releaseTimestampSupport` is used. If we set it to `true` and the value isn't available for some packages, what would happen?

bryannaegele commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+
+      return releaseResult;
+    } else {
+      // not sure how to handle error here since get or getBuffer doesn't return an error
```
I don't understand what this means. Can you point to an example or provide a suggestion?

bryannaegele commented on lib/modules/datasource/hex/index.ts:
```diff
@@ -38,31 +43,111 @@ export class HexDatasource extends Datasource {
+    // istanbul ignore else
+    if (resp.statusCode === 200) {
+      await decompressBuffer(resp.body)
+        .then((signedPackage) => {
```
Can I ask why? await -> then -> else is used extensively in the codebase. This feels like a nit.

bryannaegele commented:
**Handing to Maintainers to Finish**

I leave this PR and the Mix PR #29775 to the maintainers to complete. My limited time is better spent on my maintainer duties elsewhere.

These PRs are required to Elixir/Hex support for companies to adopt Renovate in non-trivial setups. First-order support for the hex registry is necessary to support the Hex package manager. This was unfortunately missed during original implementation in the good faith effort to support Hex/Mix.

Reasons for the handoff:
* compatibility concerns and solutions are best left to maintainers
* maintainer expertise on testing techniques employed in the project will provide better coverage
* strong opinions on code style, organization, and nits are best resolved by codeowners

Three months and >100 hours greatly surpasses the level of effort I thought this would take when I volunteered to investigate and take a stab.

Renovate is a great project. I hope you can find resources to complete the work @rarkins so we can consider using it for Elixir projects again someday.

Thanks

#29622

rarkins commented:
@bryannaegele thanks for your efforts. I've closed this if you're unable to continue and hope someone can pick it up in future.

zharinov commented:
@bryannaegele Thanks for your effort, it definitely isn't easy one 🤝 

I'll take a look if I could reuse it as series of smaller changes.
